pkg/trace/api/api.go

```diff
@@ -451,9 +456,24 @@ func (r *HTTPReceiver) handleStats(w http.ResponseWriter, req *http.Request) {
 func (r *HTTPReceiver) handleTraces(v Version, w http.ResponseWriter, req *http.Request) {
 	ts := r.tagStats(v, req.Header)
 	tracen, err := traceCount(req)
-	if err == nil && r.rateLimited(tracen) {
+	defer req.Body.Close()
+	select {
+	case r.recvsem <- struct{}{}:
+	case <-time.After(1 * time.Second):
 		// this payload can not be accepted
 		io.Copy(io.Discard, req.Body) //nolint:errcheck
+		w.WriteHeader(http.StatusRequestTimeout)
+		r.replyOK(req, v, w)
+		ts.PayloadRefused.Inc()
+		return
+	}
+
+	defer func() {
+		<-r.recvsem
+	}()
+	if err == nil && r.rateLimited(tracen) {
+		// this payload can not be accepted
+		//io.Copy(io.Discard, req.Body) //nolint:errcheck
```

Comment:
Need to revert this.
Apparently we need to read the body into discard for the keep-alive to work.